src/reason-parser/reason_parser.mly

```diff
@@ -882,34 +891,61 @@ type let_bindings =
   { lbs_bindings: Parsetree.value_binding list;
     lbs_rec: rec_flag;
     lbs_extension: (attributes * string Asttypes.loc) option;
+    lbs_bang: Longident.t Asttypes.loc option;
     lbs_loc: Location.t }
 
-let mklbs ext rf lb loc =
+let mklbs bang ext rf lb loc =
   { lbs_bindings = [lb];
     lbs_rec = rf;
     lbs_extension = ext;
+    lbs_bang = bang;
     lbs_loc = loc; }
 
 let addlbs lbs lbs' =
   { lbs with lbs_bindings = lbs.lbs_bindings @ lbs' }
 
 let val_of_let_bindings lbs =
   let str = Str.value lbs.lbs_rec lbs.lbs_bindings in
+  if lbs.lbs_bang <> None then
+    raise Syntaxerr.(Error(Not_expecting(lbs.lbs_loc, "let!foo is not allowed at the top level")));
```

Comment:
Is this just to prevent accidentally nesting everything below?

Reply:
There's nothing really that we could transform it into 🤷‍♀️  so I error out here instead of just doing nothing

Reply:
@chenglou Allowing a top-level binding to be `let!foo` would be somehow like asking for the rest of the module to go into the `foo` callback, which we can't control and doesn't make sense (because the rest of the module contains module item declarations, not only function body code).

Reply:
@aantron It's also the case that `foo` couldn't be exported. I could imagine a couple approaches to solving this though. I don't recommend continuations at the top level, but sometimes I just want to write a hacky script and get started immediately, copying and pasting from the bodies of functions so it's also worth thinking about. You could imagine some modules that define the monadic operation helping to define the meaning of monadic operations at the top level.

Reply:
We do something like this in the Lwt PPX, where a top-level `let%lwt` means wrap the stuff inside the binding in a call to `Lwt_main.run`. In the case of Lwt, though, this seems like a bad idea, as it's not compatible with Node. Nobody really knows about this behavior, and I'd like to remove it :)